src/Hmac.mo

```diff
@@ -0,0 +1,96 @@
+import Array "mo:base/Array";
+import SHA256Digest "../motoko-sha/src/SHA256";
+import SHA512Digest "../motoko-sha/src/SHA512";
+
+module {
+  public type Digest = {
+    write : ([Nat8]) -> () ;
+    sum : () -> [Nat8];
+  };
+
+  public type DigestFactory = {
+    blockSize : Nat;
+    create : () -> Digest;
+  };
+
+  public type Hmac = {
+    write : ([Nat8]) -> ();
+    sum : () -> [Nat8];
+  };
+
+  // Sha256 support.
+  object sha256DigestFactory {
+    public let blockSize : Nat = 64;
+    public func create () : Digest = SHA256Digest.Digest();
+  };
+  public func sha256(key: [Nat8]) : Hmac = HmacImpl(key, sha256DigestFactory);
+
+  // Sha512 support.
+  object sha512DigestFactory {
+    public let blockSize : Nat = 128;
+    public func create () : Digest = SHA512Digest.Digest();
+  };
+  public func sha512(key: [Nat8]) : Hmac = HmacImpl(key, sha512DigestFactory);
+
+  // Construct HMAC from an arbitrary digest function.
+  public func new(key : [Nat8], digestFactory : DigestFactory) : Hmac {
+    return HmacImpl(key, digestFactory);
+  };
+
+  // Construct HMAC from the given digest function:
+  // HMAC(key, data) = H((k' ^ outerPad) || H((k' ^ innerPad) || m))
```

Comment:
`m` should be `data`, right?

Reply:
yes, I updated the input names and forgot to change it in comments, fixed.